File: jarvis_ai/tools/webdata.py
"""Weather / news / stocks / crypto — same providers as the Electron build's webdata.ts."""

from typing import Dict, List, Optional

import requests

import config


def fetch_weather(location: Optional[str] = None) -> Dict:
    loc = (location or "").strip() or config.get_settings()["weatherLocation"]
    api_key = config.get_api_key("openweather")
    if not api_key:
        raise RuntimeError("Add an OpenWeather API key in Settings to enable live weather.")
    res = requests.get(
        "https://api.openweathermap.org/data/2.5/weather",
        params={"q": loc, "units": "metric", "appid": api_key},
        timeout=10,
    )
    if not res.ok:
        raise RuntimeError(f'Weather lookup failed ({res.status_code}) for "{loc}".')
    data = res.json()
    return {
        "location": data.get("name", loc),
        "tempC": round(data.get("main", {}).get("temp", 0)),
        "condition": (data.get("weather") or [{}])[0].get("description", "unknown"),
        "humidityPercent": data.get("main", {}).get("humidity", 0),
    }


def fetch_news() -> List[Dict]:
    api_key = config.get_api_key("newsapi")
    if not api_key:
        raise RuntimeError("Add a NewsAPI key in Settings to enable live headlines.")
    res = requests.get(
        "https://newsapi.org/v2/top-headlines",
        params={"language": "en", "pageSize": 8, "apiKey": api_key},
        timeout=10,
    )
    if not res.ok:
        raise RuntimeError(f"News lookup failed ({res.status_code}).")
    articles = res.json().get("articles", [])
    return [
        {"id": str(i), "title": a.get("title"), "source": (a.get("source") or {}).get("name", "unknown"), "url": a.get("url")}
        for i, a in enumerate(articles)
    ]


def fetch_stock_quote(symbol: str) -> str:
    res = requests.get("https://stooq.com/q/l/", params={"s": symbol, "f": "sd2t2ohlcv", "h": "", "e": "csv"}, timeout=10)
    if not res.ok:
        raise RuntimeError("Stock lookup failed.")
    lines = res.text.strip().split("\n")
    if len(lines) < 2:
        raise RuntimeError(f'No quote found for "{symbol}".')
    cols = lines[1].split(",")
    close = cols[6] if len(cols) > 6 else "N/D"
    if close == "N/D":
        raise RuntimeError(f'No quote found for "{symbol}".')
    return f"{symbol.upper()}: {close}"


def fetch_crypto_price(symbol: str) -> str:
    coin_id = symbol.lower()
    res = requests.get(
        "https://api.coingecko.com/api/v3/simple/price", params={"ids": coin_id, "vs_currencies": "usd"}, timeout=10
    )
    if not res.ok:
        raise RuntimeError("Crypto lookup failed.")
    data = res.json()
    price = data.get(coin_id, {}).get("usd")
    if price is None:
        raise RuntimeError(f'No price found for "{symbol}". Try the full CoinGecko id (e.g. "bitcoin").')
    return f"{symbol.upper()}: ${price:,}"

File: jarvis-desktop/electron/services/system/webdata.ts
import type { NewsHeadline, WeatherSnapshot } from "../../../shared/types";
import { getApiKey, getSettings } from "../../store";

export async function fetchWeather(location?: string): Promise<WeatherSnapshot> {
  const loc = location?.trim() || getSettings().weatherLocation;
  const apiKey = getApiKey("openweather");
  if (!apiKey) {
    throw new Error("Add an OpenWeather API key in Settings to enable live weather.");
  }
  const url = `https://api.openweathermap.org/data/2.5/weather?q=${encodeURIComponent(loc)}&units=metric&appid=${apiKey}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Weather lookup failed (${res.status}) for "${loc}".`);
  const data = (await res.json()) as any;
  return {
    location: data.name ?? loc,
    tempC: Math.round(data.main?.temp ?? 0),
    condition: data.weather?.[0]?.description ?? "unknown",
    humidityPercent: data.main?.humidity ?? 0,
    updatedAt: Date.now(),
  };
}

export async function fetchNews(): Promise<NewsHeadline[]> {
  const apiKey = getApiKey("newsapi");
  if (!apiKey) {
    throw new Error("Add a NewsAPI key in Settings to enable live headlines.");
  }
  const url = `https://newsapi.org/v2/top-headlines?language=en&pageSize=8&apiKey=${apiKey}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`News lookup failed (${res.status}).`);
  const data = (await res.json()) as any;
  return (data.articles ?? []).map((a: any, i: number) => ({
    id: String(i),
    title: a.title,
    source: a.source?.name ?? "unknown",
    url: a.url,
    publishedAt: new Date(a.publishedAt).getTime(),
  }));
}

export async function fetchStockQuote(symbol: string): Promise<string> {
  // Stooq offers free, keyless end-of-day/delayed quotes — good enough for a HUD readout
  // without requiring the user to provision a brokerage API key.
  const res = await fetch(`https://stooq.com/q/l/?s=${encodeURIComponent(symbol)}&f=sd2t2ohlcv&h&e=csv`);
  if (!res.ok) throw new Error("Stock lookup failed.");
  const csv = await res.text();
  const [, row] = csv.trim().split("\n");
  const cols = row?.split(",") ?? [];
  const close = cols[6];
  if (!close || close === "N/D") throw new Error(`No quote found for "${symbol}".`);
  return `${symbol.toUpperCase()}: ${close}`;
}

export async function fetchCryptoPrice(symbol: string): Promise<string> {
  const id = symbol.toLowerCase();
  const res = await fetch(`https://api.coingecko.com/api/v3/simple/price?ids=${encodeURIComponent(id)}&vs_currencies=usd`);
  if (!res.ok) throw new Error("Crypto lookup failed.");
  const data = (await res.json()) as Record<string, { usd: number }>;
  const price = data[id]?.usd;
  if (price === undefined) throw new Error(`No price found for "${symbol}". Try the full CoinGecko id (e.g. "bitcoin").`);
  return `${symbol.toUpperCase()}: $${price.toLocaleString()}`;
}
